### PR TITLE
Changing structure and moving code around, updating code generation

### PR DESCRIPTION
**Pull request by alejandroEsc**
Here I am waiting for code changes from code quality checks, and changes due to making this work with dep to ensure this works well. 

Sample of what the code looks when running:
```
$ export CLUSTER_CONTROLLER_KUBECONFIG=~/.kube/config
$ go run ./cmd/clustercontroller/main.go
2018-04-20 04:41:26 INFO cmd.clustercontroller main.go:45 Starting Cluster-Controller...
2018-04-20 04:41:26 INFO cmd.clustercontroller main.go:70 Constructing informer factory
2018-04-20 04:41:26 INFO cmd.clustercontroller main.go:73 Constructing controller
```

### CODE REVIEW

**Comment by alejandroEsc** [comment]
@leahnp no worries, its not done yet since I am waiting on two other PRs and sure I can add your PR as something I am holding off. Just let me know when its cleared.

**Comment by oneilcin** [comment]
I am glad you are testing dep with the latest code in our repo.  Good luck!  I like the deployment you have for the controller.

**Comment by alejandroEsc** [comment]
@oneilcin if you feel this is ok, please merge, thanks!

**Comment by alejandroEsc** [comment]
Ready for review. @oneilcin @leahnp @sostheim  finished moving things around, making a few things more readable, and  making sure we can do code generation regardless of machine. 

```
2018-04-24 05:24:24 INFO cmd.clustercontroller main.go:43 Starting Cluster-Controller...
2018-04-24 05:24:24 INFO cmd.clustercontroller main.go:68 Constructing informer factory
2018-04-24 05:24:24 INFO cmd.clustercontroller main.go:71 Constructing controller
```

**Comment by alejandroEsc** [comment]
by how much will it delay?

**Comment by oneilcin** [comment]
I am also regenerating the types.go I would want to test your change before using it, since you have changed the way we build. I was curious why you needed to make the changes you did, was it required for dep to work?   Looked like code cleanup to me.

**Comment by alejandroEsc** [comment]
@oneilcin i could not get code-generation to work consistently, not necessarily because of dep just cause i could not make generation work consistently in my machine, even when i pointed to the repo stored locally. So, to avoid that issue, i expect to build the generator tools if need be and just call directly for code generation. It works and i have tested it out in two code bases.

**Comment by alejandroEsc** [comment]
@oneilcin also please look at redis-operator, @venezia uses a similar pattern.

**Comment by leahnp** [comment]
LGTM - but I will defer to Cindy's decision, I don't want her to lose any momentum considering our Friday deliverable.

**Comment by joejulian** [comment]
On principal I agree with moving scripts to `scripts/`, but the kubernetes way is to have them in `hack/`. Do we want to stay with the upstream standards?

File: pkg/apis/clustercontroller/v1alpha1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func (in *ClusterInfo) DeepCopyInto(out *ClusterInfo) {
 	out.Fabric = in.Fabric
 	if in.NodePools != nil {
 		in, out := &in.NodePools, &out.NodePools
-		*out = make([]NodeProperties, len(*in))
+		*out = make([]NodeInfo, len(*in))
 		copy(*out, *in)
```
Comment (oneilcin) [inline pkg/apis/clustercontroller/v1alpha1/zz_generated.deepcopy.go]: I don't see a reason for changing these names.

File: pkg/apis/clustercontroller/v1alpha1/types.go [hunk 2]
```diff
@@ -16,45 +16,43 @@ type KrakenCluster struct {
...
-	AzureProvider CloudProviderName = "azure"
-)
-
+// CloudProviderInfo contains the details of a cloud provider
 type CloudProviderInfo struct {
-	Name        string                   `json:"name"`
-	Credentials CloudProviderCredentials `json:"credentials,omitempty"`
+	Name        string                       `json:"name"`
+	Credentials CloudProviderCredentialsInfo `json:"credentials,omitempty"`
 }
```
[the remark this answers is not in the document]
Comment (alejandroEsc) [reply pkg/apis/clustercontroller/v1alpha1/types.go]: Not sure the logic of types and constants belong together. Instantces of types should be moved out. So not sure what the logic of what you are thinking is? Can i see an example?

**Comment by alejandroEsc** [comment]
@oneilcin we arent kubernetes repo, and i havent eliminated hack directory?

**Comment by sostheim** [comment]
I have a number of thought that I'll try to briefly summarize here.  

First, I appreciate the effort that went in to this work.  Clearly there was a lot of time, thought, and effort that went in to this work.

However, this PR is too large as it is.  It addresses to many different concerns that can't all be easily reviewed within a single context.  The changes vary from the original intent of getting a consistent code generation experience.  The commits include a number of refactoring changes, to adding new functionality (e.g. viper + env vars that I can't find in the code only in the README).

My chief concern is that these are largely stylistic changes that not strictly necessary.  It is completely acceptable to have different stylistic conventions with respect to naming and structure of code.  The only real qualification we impose on code quality are the accepted best practices of running go fmt\vet\link.  

I do think we need to breakout some of this work in to separate tickets that address individual concerns.  For example, if it is a distinct change, I would like to keep the removal of the `glide` dependency management.   I would like to refine the change here down to the set of steps necessary to make the code generation changes independent (maybe easier in a separate ticket).  

After discussing this further with @alejandroEsc out of band, we'll close this PR and create new, separate ones as necessary.

**Comment by alejandroEsc** [comment]
:) happy to close it, will get to the items we discussed out of band asap!

**Comment by alejandroEsc** [comment]
One last comment. I respectfully disagree with the decision made here. For one this isnt stylistic at all. Its a choice to bring code to a standard that is acceptable in the community and that we should strive for internally. The choices I made here are not arbitrary but deliberate and merit a whole hearted review. Even if this is a POC and internal facing I would personally not be satisfied with its current state, thats my opinion, sure. 

As for the PR being too large I admit there are pieces that can be broken down into smaller constituents, but that is after the fact and this ticket was deliberately left open as a WIP so that others can chime in as I made changes and rebased my code to accommodate other's work.

**Comment by alejandroEsc** [comment]
oh and here are the things that I will bring forward:
- Add PR to include linting tests
- Add PR to enable CI to use those tests as a gate. 
- Dep works, no further work needs to be in place here.

that is all. Thanks again :)

**Pull request closed without merging**